src/AgentImplementation.sol

```diff
@@ -298,36 +293,29 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
     }
 
     function _chargeFeeByMsgValue() internal {
-        if (msg.value == 0) return;
-
-        // TODO: implement new msg.value fee charge mechanism
+        if (msg.value == 0) {
+            return;
+        } else {
+            uint256 feeRate = IRouter(router).feeRate();
+            address feeCollector = IRouter(router).feeCollector();
+            IParam.Fee memory fee = FeeLogic.getFee(_NATIVE, msg.value, feeRate, _NATIVE_FEE_META_DATA);
+            fee.charge(feeCollector);
+        }
     }
 
-    function _chargeFee(IParam.Fee[] memory fees, bool chargeFromUser) internal {
+    function _chargeFee(IParam.Fee[] memory fees, bool chargeFromAgent) internal {
         uint256 length = fees.length;
         if (length == 0) return;
 
         address feeCollector = IRouter(router).feeCollector();
         for (uint256 i; i < length; ) {
-            uint256 amount = fees[i].amount;
-            if (amount == 0) {
-                unchecked {
-                    ++i;
-                }
-                continue;
-            }
-
             address token = fees[i].token;
-            if (token == _NATIVE) {
-                payable(feeCollector).sendValue(amount);
-            } else if (chargeFromUser) {
-                address user = IRouter(router).currentUser();
-                IAllowanceTransfer(permit2).transferFrom(user, feeCollector, amount.toUint160(), token);
+            if (token == _NATIVE || chargeFromAgent) {
+                fees[i].charge(feeCollector);
             } else {
-                IERC20(token).safeTransfer(feeCollector, amount);
+                address user = IRouter(router).currentUser();
+                fees[i].chargeFrom(feeCollector, user, permit2);
```

Comment:
suggest put `user` as the first param since we are charging from `user`.

Reply:
Sounds good, in line with the ERC20 convention:
- function transfer(address to, uint256 amount) external returns (bool);
- function transferFrom(address from, address to, uint256 amount) external returns (bool);

Reply:
If so, I suggest we can change the functions to

- function pay(address feeCollector)
- function payFrom(address user, address feeCollector)

Reply:
Sounds good!